stan/math/prim/mat/fun/multiply.hpp

```diff
@@ -54,7 +57,19 @@ inline Eigen::Matrix<double, R1, C2> multiply(
     const Eigen::Matrix<double, R1, C1>& m1,
     const Eigen::Matrix<double, R2, C2>& m2) {
   check_multiplicable("multiply", "m1", m1, "m2", m2);
-  return m1 * m2;
+#ifdef STAN_OPENCL
+  if (m1.rows() * m1.cols() * m2.cols()
+      > opencl_context.tuning_opts().multiply_dim_prod_worth_transfer) {
+    matrix_cl<double> m1_cl(m1);
+    matrix_cl<double> m2_cl(m2);
+    matrix_cl<double> m3_cl = m1_cl * m2_cl;
+    return from_matrix_cl(m3_cl);
+  } else {
+#endif
+    return m1 * m2;
+#ifdef STAN_OPENCL
```

Comment:
Let's just make this one ifdef.

On a sidenote, been thinking about how to get rid of these ifdefs. A thing we could do is 

1. In the `opencl_context.hpp` have an `#ifundef STAN_OPENCL`, then define empty classes / methods
2. Define something like `#define STAN_USE_OPENCL false`
3. For if statements like the one above have `... & STAN_USE_OPENCL` at the end so the compiler sees the above will always evaluate to false

Then we can get rid of all these `#ifdef`s. Does the above make sense?

Reply:
With 1 ifdef it looks like this? Is that how you meant it. 

```cpp
#ifdef STAN_OPENCL
  if (m1.rows() * m1.cols() * m2.cols()
      > opencl_context.tuning_opts().multiply_dim_prod_worth_transfer) {
    matrix_cl<double> m1_cl(m1);
    matrix_cl<double> m2_cl(m2);
    matrix_cl<double> m3_cl = m1_cl * m2_cl;
    return from_matrix_cl(m3_cl);
  }
#endif
  return m1 * m2;
```

Yes, we really need a solution for this. On the mdivide and multiply rev branches its only more disgusting :)

Reply:
I like your solution. I can take a crack at that if you are busy or want to finish the matrix_cl rev code.

Reply:
1 ifdef like this

```cpp
#ifdef STAN_OPENCL
  if (m1.rows() * m1.cols() * m2.cols()
      > opencl_context.tuning_opts().multiply_dim_prod_worth_transfer) {
    matrix_cl<double> m1_cl(m1);
    matrix_cl<double> m2_cl(m2);
    matrix_cl<double> m3_cl = m1_cl * m2_cl;
    return from_matrix_cl(m3_cl);
  } else {
    return m1 * m2;
  }
#else
return m1 * m2;
#endif
```

duplicates code but gets ride of the ifdef. imo just prefer less ifdefs

Reply:
> I like your solution. I can take a crack at that if you are busy or want to finish the matrix_cl rev code.

Yeah if you have time that would be rad!

Reply:
Will do!